### PR TITLE
#162948998 users should be able to follow an unfollow each other

### PR DESCRIPTION
**Pull request by Hesbon5600**
**What does this PR do?**  
 _A user should be able to:_
- Follow another author
- Unfollow an author
- List all his/her followers
- View the authors he/she is following

**How should this be manually tested?**

1. Run the server:
`python manage.py runserver`

1. From postman:

    1. Follow an author by sending a `post` request to the '/api/v1/users/`<username>`/follow/' end point. The `username` is the username of the author you wish to follow.

    2. Retrieve a list of all followers of an author has by sending a `get` request to the  '/api/v1/users/`<username>`/followers/' end point
    3. Retrieve a list of all author a user is following by sending a `get` request to the  '/api/v1/users/`<username>`/following/' end point
    4. Unfollow an author by sending a `delete` request to the '/api/v1/users/`<username>`/follow/' end point. The `username` is the username of the author you wish to unfollow.
    
**Any background context you want to provide?**  
None

**What are the relevant pivotal tracker stories?**
[#162948998](https://www.pivotaltracker.com/story/show/162948998)

**Screenshots**
   1. Follow a user
screenshot 2019-01-29 at 13 31 52

   2. Retrieve all followers of a user
screenshot 2019-01-29 at 13 32 08

   3. Retrieve all users an author is following
screenshot 2019-01-29 at 13 32 15

   4. Unfollow an author
screenshot 2019-01-29 at 13 32 40

### CODE REVIEW

File: authors/apps/profiles/views.py [hunk 1]
```diff
@@ -1,14 +1,36 @@
...
 
+def current_user_profile(request):
+    authenticated_user = request.user
+    return authenticated_user.profile
+
+
+def followed_user_profile(username):
+    profile_to_follow = Profile.objects.get(user__username=username)
+    return profile_to_follow
+
```
Comment (Nyakaru) [inline authors/apps/profiles/views.py:28]: Good use of flake8 on formatting your files.

**Review by Nyakaru — COMMENTED** [review]
Kindly add some tests to fix coveralls

**Comment by Nicanor008** [comment]
Spelling error on PR How this should be tested 2(i) auhor

**Review by charitymarani — COMMENTED** [review]
Add tests to increase test coverage

File: authors/apps/profiles/views.py [hunk 2]
```diff
@@ -46,3 +68,101 @@ def update(self, request, username):
...
+    def get(self, request, username):
+        """Following"""
+
+        following_user = current_user_profile(request)
+        try:
+            followed_user = followed_user_profile(username)
+        except Exception:
+            raise NotFound(
+                'No such user found,please check the username again')
```
Comment (charitymarani) [inline authors/apps/profiles/views.py:150]: Don't just test the status code,Test these messages too to increase test coverage

**Comment by eduhmik** [comment]
Hey @Hesbon5600 there's a mix match in your screenshots

File: authors/apps/profiles/tests/test_following.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        self.forbidden_403(response)
+
+    def test_get_following_with_auth(self):
+        token = self.login_user2()
+        response = self.client.get(
+            self.followers_following_url(),
+            HTTP_AUTHORIZATION='Token ' + token, format='json')
+        self.ok_200(response)
+
+    def test_get_followers_with_auth(self):
```
Comment (JoshuaKodhe) [inline authors/apps/profiles/tests/test_following.py:90]: This test contains repetition from the test on line 56, you can refactor to make your code DRY

**Comment by Nyakaru** [comment]
@Hesbon5600 Also check the typo in your PR title

File: authors/apps/profiles/tests/test_following.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        response = self.client.delete(
+            self.follow_existing_user_url(), format='json')
+        self.not_found_404(response)
+        msesage = 'You must be logged in first'
+        self.response_message_test(msesage, response)
+
+    def test_follow_non_existing_user(self):
+        response = self.follow(self.follow_non_existing_user_url())
+        self.not_found_404(response)
+        msesage = 'You are trying to follow a user who doe not exist, please check the username again'
```
Comment (kzyangiro) [inline authors/apps/profiles/tests/test_following.py]: @Hesbon5600 "Doe not exist" A typo here.

**Review by kzyangiro — CHANGES_REQUESTED** [review]
@Hesbon5600 Good job, You could however add more description to your outputs since as at now, all your output is the same, unless the user looks at the url or the request method. Show success messages and an info name for the data displayed:

"message": " Successfully unfollowed Hesbon" ,
"Hesbon's info": {
    "username": "hesbon",
    "followers_count": 6
    ----
}

File: authors/apps/profiles/tests/test_following.py (hunk 4 above)
Comment (Hesbon5600) [reply authors/apps/profiles/tests/test_following.py]: Fixed

**Comment by Hesbon5600** [comment]
> @Hesbon5600 Good job, You could however add more description to your outputs since as at now, all your output is the same, unless the user looks at the url or the request method. Show success messages and an info name for the data displayed:
> 
> "message": " Successfully unfollowed Hesbon" ,
> "Hesbon's info": {
> "username": "hesbon",
> "followers_count": 6
> ----
> }

Implemented

**Review by kzyangiro — APPROVED** [review]

**Pull request merged**